include/envoy/config/ads.h

```diff
@@ -0,0 +1,69 @@
+#pragma once
+
+#include "envoy/common/exception.h"
+#include "envoy/common/pure.h"
+
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Config {
+
+class AdsCallbacks {
+public:
+  virtual ~AdsCallbacks() {}
+
+  /**
+   * Called when a configuration update is received.
+   * @param resources vector of fetched resources corresponding to the configuration update.
+   * @throw EnvoyException with reason if the configuration is rejected. Otherwise the configuration
+   *        is accepted. Accepted configurations have their version_info reflected in subsequent
+   *        requests.
+   */
+  virtual void onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources) PURE;
+
+  /**
+   * Called when either the subscription is unable to fetch a config update or when onConfigUpdate
+   * invokes an exception.
+   * @param e supplies any exception data on why the fetch failed. May be nullptr.
+   */
+  virtual void onConfigUpdateFailed(const EnvoyException* e) PURE;
+};
+
+/**
+ * Handle on an ADS subscription that enables the subscription to be canceled.
+ */
+class AdsWatch {
+public:
+  virtual ~AdsWatch() {}
+
+  /**
+   * Cancel an ADS watch that was created via AdsApi::subscribe. This will free the AdsWatch object.
+   */
+  virtual void cancel() PURE;
```

Comment:
Any particular reason to not have subscribe() return a unique_ptr and destruction of handle deletes watch? Not a big deal, just curious. Slightly easier for the caller.

Reply:
Sure, good idea.